tests/30rooms/12thirdpartyinvite.pl

```diff
@@ -0,0 +1,419 @@
+use Crypt::NaCl::Sodium;
+use Protocol::Matrix qw( encode_json_for_signing encode_base64_unpadded );
+
+my $crypto_sign = Crypt::NaCl::Sodium->sign;
+
+test "Can invite existing 3pid",
+   requires => [qw( user more_users test_http_server_hostandport )],
+
+   do => sub {
+      my ( $inviter, $more_users, $id_server ) = @_;
+
+      my $invitee_email = "marmosets\@monkeyworld.org";
+      my $invitee_mxid = $more_users->[0]->user_id;
+      my $room_id;
+
+      Future->needs_all(
+         stub_is_lookup( $invitee_email, $invitee_mxid ),
+
+         matrix_create_and_join_room( [ $inviter ], visibility => "private" )
+         ->then( sub {
+            ( $room_id ) = @_;
+            do_request_json_for( $inviter,
+               method => "POST",
+               uri    => "/api/v1/rooms/$room_id/invite",
+
+               content => {
+                  id_server => $id_server,
+                  medium => "email",
+                  address => $invitee_email,
+                  display_name => "Cute things",
+               },
+            ),
+         })->then( sub {
+            matrix_get_room_state( $inviter, $room_id,
+               type => "m.room.member",
+               state_key => $invitee_mxid,
+            )->on_done( sub {
+               my ( $body ) = @_;
+
+               log_if_fail "Body", $body;
+               $body->{membership} eq "invite" or
+                  die "Expected invited user membership to be 'invite'";
+            });
+         }),
+      );
+   };
+
+test "Can invite unbound 3pid",
+   requires => [qw( user more_users test_http_server_hostandport first_home_server )],
+   do => \&can_invite_unbound_3pid;
+
+test "Can invite unbound 3pid over federation",
+   requires => [qw( user remote_users test_http_server_hostandport first_home_server )],
+   do => \&can_invite_unbound_3pid;
+
+sub can_invite_unbound_3pid {
+   my ( $inviter, $other_users, $id_server, $user_agent ) = @_;
+   my $invitee = $other_users->[0];
+
+   make_3pid_invite( $inviter, $invitee, $id_server, 1, sub {
+      my ( $token, $public_key, $signature, $room_id ) = @_;
+
+      do_request_json_for( $invitee,
+         method => "POST",
+         uri    => "/api/v1/rooms/$room_id/join",
+         content => {
+            token => $token,
+            public_key => encode_base64_unpadded( $public_key ),
+            signature => $signature,
+            key_validity_url => "https://$id_server/_matrix/identity/api/v1/pubkey/isvalid",
```

Comment:
Align the fat commas (`=>`) with leading whitespace before; so they all line up.